Source/Dafny/Compilers/Compiler-python.cs

```diff
@@ -233,7 +260,66 @@ protected class ClassWriter : IClassWriter {
     protected override string TypeInitializationValue(Type type, ConcreteSyntaxTree wr, IToken tok,
       bool usePlaceboValue,
       bool constructTypeParameterDefaultsFromTypeDescriptors) {
-      throw new NotImplementedException();
+      var xType = type.NormalizeExpandKeepConstraints();
+
+      if (usePlaceboValue) {
+        return "undefined";
+      }
+
+      if (xType is BoolType) {
+        return "false";
+      } else if (xType is CharType) {
+        return CharType.DefaultValueAsString;
+      } else if (xType is IntType || xType is BigOrdinalType) {
+        return IntegerLiteral(0);
+      }
+
+      var udt = (UserDefinedType)xType;
+      var cl = udt.ResolvedClass;
+      Contract.Assert(cl != null);
+      if (cl is SubsetTypeDecl) {
+        var td = (SubsetTypeDecl)cl;
+        if (td.WitnessKind == SubsetTypeDecl.WKind.Compiled) {
+          return TypeName_UDT(FullTypeName(udt), udt, wr, udt.tok) + ".Default";
+        } else if (td.WitnessKind == SubsetTypeDecl.WKind.Special) {
+          // WKind.Special is only used with -->, ->, and non-null types:
+          Contract.Assert(ArrowType.IsPartialArrowTypeName(td.Name) || ArrowType.IsTotalArrowTypeName(td.Name) || td is NonNullTypeDecl);
+          if (ArrowType.IsPartialArrowTypeName(td.Name)) {
+            return "null";
+          } else if (((NonNullTypeDecl)td).Class is ArrayClassDecl) {
+            // non-null array type; we know how to initialize them
+            var arrayClass = (ArrayClassDecl)((NonNullTypeDecl)td).Class;
+            if (arrayClass.Dims == 1) {
+              return "[]";
+            } else {
+              return string.Format("_dafny.newArray(undefined, {0})", Util.Comma(arrayClass.Dims, _ => "0"));
+            }
+          } else {
+            // non-null (non-array) type
+            // even though the type doesn't necessarily have a known initializer, it could be that the the compiler needs to
+            // lay down some bits to please the C#'s compiler's different definite-assignment rules.
+            return "null";
+          }
+        } else {
+          return TypeInitializationValue(td.RhsWithArgument(udt.TypeArgs), wr, tok, usePlaceboValue, constructTypeParameterDefaultsFromTypeDescriptors);
+        }
+      } else {
+        Contract.Assert(false); throw new cce.UnreachableException();  // unexpected type
+      }
+
+    }
+
+    static string FormatDefaultTypeParameterValue(TopLevelDecl tp) {
+      Contract.Requires(tp is TypeParameter || tp is OpaqueTypeDecl);
+      return $"_default_{tp.CompileName}";
+    }
+    string IntegerLiteral(BigInteger i) {
+      if (i.IsZero || i.IsOne) {
```

Comment:
Why `i.isOne` would result in `int()` ? I thought `int()` always returned zero.

Reply:
made a mistake there, thanks. Removing the `i.IsOne` from if condition so that it will go through the else condition.